packages/xsnap/test/xsnap-eof.test.js

```diff
@@ -0,0 +1,268 @@
+            process.kill(process.pid, 'SIGKILL');
+            return new Uint8Array();
+        }
+
+        const vat = await xsnap({ ...options(io), handleCommand });
+        await vat.evaluate('function handleCommand(message) { issueCommand(new Uint8Array().buffer); };');
+        await vat.issueStringCommand('0');
+        console.error('Error: parent was not killed!');
+        await vat.close();
+    `;
+
+  let nodeExited = false;
+  let nodeError = null;
+  let nodeExitCode = null;
+  let nodeExitSignal = null;
+  let strOut = '';
+  let strErr = '';
+
+  const p = proc.spawn('node', ['--input-type=module', '-e', parentText], {
+    cwd: `${process.cwd()}/test`,
+  });
+
+  p.stdout.on('data', data => {
+    strOut = `${strOut}${data}`;
+  });
+
+  p.stderr.on('data', data => {
+    strErr = `${strErr}${data}`;
+  });
+
+  p.on('error', err => {
+    nodeError = err;
+  });
+
+  p.on('exit', (code, signal) => {
+    nodeExited = true;
+    nodeExitCode = code;
+    nodeExitSignal = signal;
+  });
+
+  p.stdin.end();
+
+  await waitFor(() => {
+    return nodeExited;
+  });
+
+  t.is(`${strOut}`, ``, 'wrong stdout output');
+  t.regex(`${strErr}`, /Has parent died\?/, 'wrong stderr output');
+  t.is(nodeError, null, 'parent returned an error');
+  t.is(nodeExited, true, 'parent did not exit');
+  t.is(nodeExitCode, null, 'parent exit code is not null');
+  t.is(`${nodeExitSignal}`, `SIGKILL`, 'wrong parent exit signal');
+});
+
+async function doSetup(handleCommand) {
+  let xsnapProcess;
+  const spawnSpy = (...args) => {
+    xsnapProcess = proc.spawn(...args);
+    return xsnapProcess;
+  };
+
+  const io = { spawn: spawnSpy, os: os.type(), fs, tmpName };
+  const vat = await xsnap({ ...options(io), handleCommand, stderr: 'pipe' });
+  await vat.evaluate(
+    'function handleCommand(message) { issueCommand(new Uint8Array().buffer); };',
+  );
+  return { vat, xsnapProcess };
+}
+
+test('xsnap-worker complains while trying to READ an answer when pipes are closed', async t => {
+  let xsnapProcess;
+  let toXsnap;
+  let fromXsnap;
+  let issueCommandError;
+  let vatCloseError;
+  let vatExitCode;
+  let vatExitSignal;
+  let strErr = '';
+
+  async function handleCommand(_message) {
+    // suspend the child process so the pipes appear
+    // to close atomically from its perspective
+    xsnapProcess.kill('SIGSTOP');
+    // wait for signal delivery
+    await waitFor(() => {
+      return xsnapProcess.killed;
+    });
+
+    // close the pipes
+    toXsnap.end();
+    toXsnap.destroy();
+    fromXsnap.end();
+    fromXsnap.destroy();
+
+    // un-suspend the child process
+    xsnapProcess.kill('SIGCONT');
+    // wait for signal delivery
+    await waitFor(() => {
+      return xsnapProcess.killed;
+    });
+
+    return new Uint8Array();
+  }
+
+  const setup = await doSetup(handleCommand);
+  const { vat } = setup;
+  ({ xsnapProcess } = setup);
+
+  toXsnap = xsnapProcess.stdio[3];
```

Comment:
`test/xsnap-eof.test.js(142,13): error TS18048: 'xsnapProcess' is possibly 'undefined'`

Well, it really could be, but that would also mean doSetup() failed, which invalidates the whole test any way. What is the preferred solution for a case like that?

Reply:
Yeah this one is particularly annoying to deal with. Since we're disconnected from the `spawn`, the types don't know which args were passed, and so what the specific type of `stdio` is. Easiest here is a cast

```suggestion
  toXsnap = /** @type {import('stream').Readable} */(xsnapProcess.stdio[3]);
```

Reply:
`/** @type {import('stream').Readable} */ `does not really help since it complains about `xsnapProcess` possibly being undefined, since it cannot track the execution flow in which assignment logic happens...